core-rust/state-manager/src/state_manager.rs

```diff
@@ -287,73 +287,28 @@ where
-        //
-        // TODO - Remove when this check lives in the engine/executor.
-        //========================================================================================
-
-        let prepared_proposed_results: Vec<_> = prepare_request
-            .proposed_transactions
-            .iter()
-            .map(
-                |raw| -> Result<
-                    (RawLedgerTransaction, PreparedLedgerTransaction),
-                    TransactionValidationError,
-                > {
-                    let ledger_transaction = LedgerTransaction::from_raw_user(raw)
-                        .map_err(|err| {
-                            TransactionValidationError::PrepareError(PrepareError::DecodeError(err))
-                        })?
-                        .to_raw()?;
-                    let prepared = self
-                        .ledger_transaction_validator
-                        .prepare_from_raw(&ledger_transaction)?;
-                    Ok((ledger_transaction, prepared))
-                },
-            )
-            .collect();
-
-        let mut intent_hash_potential_conflicts =
-            HashMap::<IntentHash, IntentHashDuplicateWith>::new();
-
-        for (_, prepared) in prepared_proposed_results.iter().flatten() {
-            let intent_hash = prepared
-                .as_user()
-                .expect("Proposed was created from user")
-                .intent_hash();
-            if read_store
-                .get_txn_state_version_by_identifier(&intent_hash)
-                .is_some()
-            {
-                intent_hash_potential_conflicts
-                    .insert(intent_hash, IntentHashDuplicateWith::Committed);
-            }
-        }
-
-        //========================================================================================
-        // PART 2:
-        // We ensure all the ancestor transactions are in our execution cache, ready to execute
-        // the ancestor transactions
+        // We execute all the ancestor transactions (on a happy path: only making sure they are in
+        // our execution cache),
         //========================================================================================
 
         let pending_transaction_base_state = AtState::PendingPreparingVertices {
             base_committed_state_version: base_state_version,
         };
 
         let mut state_tracker = StateTracker::initial(base_state_version, base_ledger_hashes);
+        let mut duplicate_intent_hash_detector =
+            DuplicateIntentHashDetector::new(read_store.deref());
 
         for raw_ancestor in prepare_request.prepared_uncommitted_transactions {
-            // TODO - By passing through the accumulator / tree hash, avoid doing all this validation if the
-            // transactions are already in the execution cache!
+            // TODO(optimization-only): We could avoid the validations here (mostly: creation of
```

Comment:
The main savings I imagine are hashing, decoding and signature verification, rather than creating the executable itself. But your point stands :)